### ISSUE TITLE
tox21: wrong handling of missing values + benchmarking

### ISSUE THREAD

**Issue by eguidotti** [deepchem/deepchem#2296]
Hi there,
first, thanks for putting together this great project!

I'm working on a new machine learning algorithm and benchmarking against the tox21 dataset.
Unfortunately I see some inconsistencies regarding missing values. From this [link](https://deepchemdata.s3-us-west-1.amazonaws.com/datasets/tox21.csv.gz) I'm able to download the csv that contains the data in the form: `0,0,1,,,0,0,1,0,0,0,0,TOX3021,CCOc1ccc2nc(S(N)(=O)=O)sc2c1`
However, when I load the data via the [API](https://deepchem.readthedocs.io/en/latest/api_reference/moleculenet.html#tox21-datasets) the missing values are treated as zeros. This is an issue as all the missing values are attributed to the class `0`, which is a quite different thing. I'm afraid this is also impacting all the evaluations of the models [here](https://github.com/deepchem/deepchem/tree/master/examples/tox21).

A second problem concerns the train/validation/test split. With the recommended split I get 6264 observations in the training set, 783 in the validation set, and 784 in the test set. However from [this paper](https://www.frontiersin.org/articles/10.3389/fenvs.2015.00080/full) the tox21 dataset consisted of 11764 training samples, 296 validation, and 647 test. 

> This dataset consisted of a training dataset of 11,764, a leaderboard set of 296, and a test set of 647 compounds

This makes benchmarking impossible with the original papers and later studies such as:
- https://www.frontiersin.org/articles/10.3389/fenvs.2015.00080/full
- https://arxiv.org/pdf/1706.02515v5.pdf

Is it possible to fix the load API with `NaN` for missing values?
Is there some option to import the original dataset and corresponding train/validation/test sets?
Many thanks!

**Comment by rbharath** [comment]
The missing data is attributed to class 0, but the weights are set to 0 as well which means we are fine on the evaluation :). This is a tricky point and it's not made clear in the documentation so we should do better on that on our end!

Your point about the source dataset not matching the dataset here is spot on though! We need to fix this. We're working on overhauling the MoleculeNet benchmark suite, and we'd be glad to add the original dataset so the benchmarking is more comparable with the source website

**Comment by eguidotti** [comment]
Hello @rbharath, thank you for your reply.
Yes it would be great to have the original dataset.

Regarding the missing values, maybe I'm missing something, but it seems to me that e.g. [here](https://github.com/deepchem/deepchem/blob/master/examples/tox21/tox21_logreg.py) there is no reference to weights or filtering of missing values. Moreover, the values would have an impact on training as well as evaluation. Am I missing something?
In any case, I'd suggest to import missing values as `NaN`. For most users:
- it would be crystal clear that the value is missing and not of the given class 
- it would make sure that models are not wrongly trained and evaluated on data treating missing values with some other class
- it would be more transparent and force the user to understand how to handle missing values depending on the application

I'm not sure if this missing value issue is related to the tox21 dataset only, or is a more general issue affecting other datasets.
Thank you again for your efforts!

**Comment by peastman** [comment]
> it seems to me that e.g. here there is no reference to weights or filtering of missing values.

The model automatically multiplies the losses by the weights during fitting and evaluation, so any datapoint with a weight of 0 is ignored.  This is discussed in some of the tutorials.  For example, see https://github.com/deepchem/deepchem/blob/master/examples/tutorials/04_Molecular_Fingerprints.ipynb.

> In any case, I'd suggest to import missing values as NaN.

That wouldn't work, because 0*NaN = NaN.  Once you allow a NaN anywhere into the computation, it's pretty much guaranteed the output will be NaN as well.  Multiplying by a zero weight doesn't help.

**Comment by eguidotti** [comment]
I see, thank you for the tutorial. 
The approach seems a bit tricky yes and at least I'd replace `0` with e.g. `999999` to identify missing values and avoid confusion.
I think I'm going to use the raw dataset from tox21 challenge, that would be easier. Thanks
